### PR TITLE
Remove state from memory by writing to database

### PR DESCRIPTION
**Pull request by josephroberts**
Change-type: minor
Signed-off-by: Joseph Roberts <joe@resin.io>

Connects to #145

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -20,15 +20,15 @@ RUN apt-get update && apt-get install -yq --no-install-recommends \
...
-RUN TAG=$(curl https://api.github.com/repos/resin-io/edge-node-manager/releases/latest -s | jq .tag_name -r) && \
-    echo "Pulling $TAG of the edge-node-manager binary" && \
-    curl -k -O https://resin-production-downloads.s3.amazonaws.com/edge-node-manager/$TAG/edge-node-manager-$TAG-linux-arm && \
-    mv edge-node-manager-$TAG-linux-arm edge-node-manager && \
-    chmod +x edge-node-manager
+# RUN TAG=$(curl https://api.github.com/repos/resin-io/edge-node-manager/releases/latest -s | jq .tag_name -r) && \
+#     echo "Pulling $TAG of the edge-node-manager binary" && \
+#     curl -k -O https://resin-production-downloads.s3.amazonaws.com/edge-node-manager/$TAG/edge-node-manager-$TAG-linux-arm && \
+#     mv edge-node-manager-$TAG-linux-arm edge-node-manager && \
+#     chmod +x edge-node-manager
```
Comment (CameronDiver) [inline Dockerfile]: Commented out code shouldn't be kept in if it's not needed anymore.

File: api/handlers.go [hunk 2]
```diff
@@ -158,11 +98,72 @@ func GetStatus(w http.ResponseWriter, r *http.Request) {
 	}
 
 	w.Header().Set("Content-Type", "application/json")
-	w.Write(bytes)
+	if _, err := w.Write(bytes); err != nil {
```
Comment (CameronDiver) [inline api/handlers.go]: I think we should check that the number of bytes written matches with the amount we want to send. If not you should attempt to resend the slice which was not sent.

File: api/handlers.go [hunk 3]
```diff
@@ -158,11 +98,72 @@ func GetStatus(w http.ResponseWriter, r *http.Request) {
...
+		"Curent status": process.CurrentStatus,
 	}).Debug("Get status")
 }
+
+func setField(w http.ResponseWriter, r *http.Request, key string, value interface{}) error {
+	vars := mux.Vars(r)
+	deviceUUID := vars["uuid"]
+
+	db, err := storm.Open(config.GetDbPath())
+	defer db.Close()
```
Comment (CameronDiver) [inline api/handlers.go]: Will this throw if the database was not able to be opened in the first place?

File: device/device.go [hunk 4]
```diff
@@ -4,137 +4,107 @@ import (
...
-	if err != nil {
-		return nil, err
+	switch d.BoardType {
+	case board.MICROBIT:
+		d.Board = microbit.Microbit{
+			Log: log,
+			Micro: nrf51822.Nrf51822{
+				Log:                 log,
+				LocalUUID:           d.LocalUUID,
+				Firmware:            nrf51822.FIRMWARE{},
```
Comment (CameronDiver) [inline device/device.go]: Should this not be microbit firmware?

File: main.go [hunk 5]
```diff
@@ -3,66 +3,78 @@ package main
...
+	"github.com/resin-io/edge-node-manager/device"
 	"github.com/resin-io/edge-node-manager/process"
+	"github.com/resin-io/edge-node-manager/radio/bluetooth"
 	"github.com/resin-io/edge-node-manager/supervisor"
 )
 
-// This variable will be populated at build time with the current version tag
-var version string
+var (
+	version   string // This variable will be populated at build time with the current version tag
```
Comment (CameronDiver) [inline main.go]: Completely preference and a massive nitpick but I feel like this comment should go above the version var line.

File: main.go [hunk 6]
```diff
@@ -3,66 +3,78 @@ package main
...
+	loopDelay time.Duration
```
Comment (CameronDiver) [inline main.go]: I would also add a little description of what this variable is for, as it's not a local.

File: main.go [hunk 7]
```diff
@@ -3,66 +3,78 @@ package main
...
+	dbDir := config.GetDbDir()
+	if err := os.MkdirAll(dbDir, os.ModePerm); err != nil {
+		log.WithFields(log.Fields{
+			"Directory": dbDir,
+			"Error":     err,
+		}).Fatal("Unable to create database directory")
+	}
+
+	db, err := storm.Open(config.GetDbPath())
+	defer db.Close()
```
Comment (CameronDiver) [inline main.go]: Same here, will this throw if we can't open the db connection?

File: main.go [hunk 8]
```diff
@@ -71,7 +83,7 @@ func init() {
 			"Port": port,
 		}).Debug("Initialising incoming supervisor API")
 
-		if err := http.ListenAndServe(port, router); err != nil {
+		if err = http.ListenAndServe(port, router); err != nil {
```
Comment (CameronDiver) [inline main.go]: Is this a typo?

File: process/process.go [hunk 9]
```diff
@@ -1,145 +1,187 @@
...
 )
 
-// Run processes the application, checking for new commits, provisioning and updating devices
-func Run(a *application.Application) []error {
-	log.Info("----------------------------------------------------------------------------------------------------")
-
-	// Put all provisioned devices associated with this application
-	defer a.PutDevices()
+func Run(a application.Application) []error {
+	log.Info("----------------------------------------")
```
Comment (CameronDiver) [inline process/process.go]: Is this debugging that's been left in?

File: process/process.go [hunk 10]
```diff
@@ -1,145 +1,187 @@
...
-			if d.Commit != d.TargetCommit && d.Status != deviceStatus.OFFLINE {
-				UpdatesPending = true
-				return
+		if errs := sendState(value); errs != nil {
+			return errs
+		}
+	}
+
+	// Refesh all provisioned devices associated with this application
+	provisionedDevices, err = getProvisionedDevices(a)
```
Comment (CameronDiver) [inline process/process.go]: I'm assuming we do need multiple calls to `getProvisionedDevices()` here and cannot cache the result?

File: process/process.go [hunk 11]
```diff
@@ -151,29 +193,185 @@ func pause() error {
...
+
+	d.Status = deviceStatus.INSTALLING
+	if err := updateDevice(d); err != nil {
+		return []error{err}
+	}
+	if errs := sendState(d); errs != nil {
+		return errs
+	}
+
+	for i := 1; i <= 3; i++ {
```
Comment (CameronDiver) [inline process/process.go]: You should export these magic numbers to a const variable, or a default function param.

File: radio/bluetooth/bluetooth.go [hunk 12]
```diff
@@ -141,7 +144,8 @@ func Scan(id string) (map[string]bool, error) {
 				return
 			case adv := <-advChannel:
 				if strings.EqualFold(adv.LocalName(), id) {
-					devices[adv.Address().String()] = true
+					var s struct{}
+					devices[adv.Address().String()] = s
```
Comment (CameronDiver) [inline radio/bluetooth/bluetooth.go]: How come we're using an empty struct as a placeholder here?

File: device/device.go (hunk 4 above)
Comment (josephroberts) [inline device/device.go]: Nope, because we have 2 `board` types: `microbit` and `nrf51822dk` and these both use the same `micro` internally: `nrf51822`

File: process/process.go (hunk 9 above)
Comment (josephroberts) [inline process/process.go]: Nope, its there to highlight the change between each managed application in the logs. It's quite hard to parse without it. For example:

With the dashes:
```
23.03.17 10:48:48 (+0000) INFO ----------------------------------------
23.03.17 10:48:48 (+0000) INFO Processing application                        Application=test1
23.03.17 10:48:58 (+0000) INFO Processing application                        Number of online devices=0
23.03.17 10:48:58 (+0000) INFO Processing application                        Number of provisioned devices=2
23.03.17 10:49:12 (+0000) INFO ----------------------------------------
23.03.17 10:49:12 (+0000) INFO Processing application                        Application=test2
23.03.17 10:49:22 (+0000) INFO Processing application                        Number of online devices=0
23.03.17 10:49:22 (+0000) INFO Processing application                        Number of provisioned devices=2
```

Without the dashes:
```
23.03.17 10:48:48 (+0000) INFO Processing application                        Application=test1
23.03.17 10:48:58 (+0000) INFO Processing application                        Number of online devices=0
23.03.17 10:48:58 (+0000) INFO Processing application                        Number of provisioned devices=2
23.03.17 10:49:12 (+0000) INFO Processing application                        Application=test2
23.03.17 10:49:22 (+0000) INFO Processing application                        Number of online devices=0
23.03.17 10:49:22 (+0000) INFO Processing application                        Number of provisioned devices=2
```

I am open to other suggestions

File: process/process.go (hunk 10 above)
Comment (josephroberts) [inline process/process.go]: The aim with writing to the database every step is avoid data loss if the container or device is restarted. The aim with reading from the database every step is to ensure I pick up changes written to the database by the `api/handlers` as soon as possible.

File: radio/bluetooth/bluetooth.go (hunk 12 above)
Comment (josephroberts) [inline radio/bluetooth/bluetooth.go]: I use a map to store the device addresses so I can easily check whether the key exists, I don't care about the value so I use a `struct{}` as it takes no memory, compared with a bool which does.
